### ISSUE TITLE
bundler 2.4.1 resolves unexpected version

### ISSUE THREAD

**Issue by doudou** [rubygems/rubygems#6226]
### Describe the problem as clearly as you can

This is a follow-up to https://github.com/rubygems/rubygems/issues/6210 (ping @deivid-rodriguez). The reproduction test-case still applies.

With ruby 2.7.5p203 (2021-11-24 revision f69aeb8314) [x86_64-linux-gnu] (Ubuntu 20.04 with brightbox ruby packages) and bundler 2.4.1, the following Gemfile

~~~
source "https://rubygems.org"
ruby "2.7.5" if respond_to?(:ruby)
gem "autoproj", ">= 2.0.0"
gem "utilrb", ">= 3.0.1"
~~~

Resolves to autoproj 2.0.3 and autobuild 1.10.0.rc3. Bundler 2.3.6 (and many versions before it) resolved to the latest versions of both packages (2.16.0, 1.24.0)

### Did you try upgrading rubygems & bundler?

Bundler is latest. Rubygems is as-packages with Ruby 2.7.5

### Post steps to reproduce the problem

https://github.com/rubygems/rubygems/issues/6210#issuecomment-1367360624 contains a Vagrantfile to install 20.04 and run shell scripts that reproduce the problem. As I mentioned to @deivid-rodriguez, these are simple shell scripts, that could be run in a docker container as well if you're not into vagrant.

### Which command did you run?

`GEM_HOME=$HOME/gems bundle install` 

### What were you expecting to happen?

Install the latest autoproj and autobuild versions (2.16.0 and 1.24.0)

### What actually happened?

Installed a very old version autoproj that depends on a pre-release of autobuild (2.0.3 and 1.10.0.rc3)

### If not included with the output of your command, run `bundle env` and paste the output below

vagrant@ubuntu2004:~$ bundle env
## Environment

```
Bundler       2.4.2
  Platforms   ruby, x86_64-linux
Ruby          2.7.5p203 (2021-11-24 revision f69aeb83146be640995753667fdd6c6f157527f5) [x86_64-linux]
  Full Path   /usr/bin/ruby2.7
  Config Dir  /etc
RubyGems      3.1.6
  Gem Home    /var/lib/gems/2.7.0
  Gem Path    /home/vagrant/.gem/ruby/2.7.0:/var/lib/gems/2.7.0:/usr/lib/ruby/gems/2.7.0:/usr/share/rubygems-integration/2.7.0:/usr/share/rubygems-integration/all:/usr/lib/x86_64-linux-gnu/rubygems-integration/2.7.0
  User Home   /home/vagrant
  User Path   /home/vagrant/.gem/ruby/2.7.0
  Bin Dir     /usr/local/bin
Tools         
  Git         2.25.1
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2023-01-01
Git SHA           2cd3ed45bf
Released Version  true
```

### Gemfile

```ruby
source "https://rubygems.org"
ruby "2.7.5" if respond_to?(:ruby)
gem "autoproj", ">= 2.0.0"
gem "utilrb", ">= 3.0.1"
```

### Gemfile.lock

```
GEM
  remote: https://rubygems.org/
  specs:
    autobuild (1.10.0.b3)
      highline
      rake (>= 0.9.0)
      utilrb (>= 1.6.0)
    autoproj (2.0.3)
      autobuild (>= 1.10.0.a)
      bundler
      concurrent-ruby
      thor (~> 0.19.0, >= 0.19.1)
      tty-color (~> 0.3.0, >= 0.3.0)
      tty-prompt
      utilrb (>= 3.0)
    concurrent-ruby (1.1.10)
    equatable (0.5.0)
    facets (3.1.0)
    highline (2.1.0)
    necromancer (0.3.0)
    pastel (0.6.1)
      equatable (~> 0.5.0)
      tty-color (~> 0.3.0)
    rake (13.0.6)
    thor (0.19.4)
    tty-color (0.3.0)
    tty-cursor (0.3.0)
    tty-prompt (0.9.0)
      necromancer (~> 0.3.0)
      pastel (~> 0.6.0)
      tty-cursor (~> 0.3.0)
      wisper (~> 1.6.1)
    utilrb (3.1.0)
      facets (>= 2.4.0)
      rake (>= 0.9)
    wisper (1.6.1)

PLATFORMS
  x86_64-linux

DEPENDENCIES
  autoproj (>= 2.0.0)
  utilrb (>= 3.0.1)

RUBY VERSION
   ruby 2.7.5p203

BUNDLED WITH
   2.4.2
```

**Comment by deivid-rodriguez** [comment]
Thanks for reporting @doudou. I had noticed this while I was working on #6210, and figured this would be a problem. I was planning to investigate but didn't find time yet, so thanks for reporting so I don't forget about this :)

In general, it's hard to always guarantee that users will get the resolution they expect because there are usually many different equally valid solutions to the version solving problem. For example, in this case Bundler 2.4 is resolving to an older version of autoproj than Bundler 2.3.26, but a newer version of utilrb than Bundler 2.3.26. And those are your two top level dependencies, which is what resolvers try to maximize.

That said:

* PubGrub (what Bundler 2.4 uses) focuses on compatibility with Molinillo (what Bundler 2.3 uses), so worth looking at trying to give the same result here.
* Normally, resolutions without pre-releases should be preferred unless really necessary, so from that perspective, Molinillo's result is superior here.

So, issue acknowledged, I will try to have a look at this soon and see what we can do here. In the mean time, feel free to change the requirement on `autoproj` to better hint our resolver on what the expected result is. For example,

```diff
diff --git a/Gemfile b/Gemfile
index f1989c5..93afc14 100644
--- a/Gemfile
+++ b/Gemfile
@@ -1,4 +1,4 @@
 source "https://rubygems.org"
 ruby "2.7.5" if respond_to?(:ruby)
-gem "autoproj", ">= 2.0.0"
+gem "autoproj", ">= 2.16.0"
 gem "utilrb", ">= 3.0.1"
```

**Comment by doudou** [comment]
> Normally, resolutions without pre-releases should be preferred unless really necessary, so from that perspective, Molinillo's result is superior here.

And vastly so. That bundler would install a pre-release version without it being explicitly selected is scary. By definition, pre-release versions are buggy.

**Comment by deivid-rodriguez** [comment]
Sorry to have scared you.

We try hard to give a proper result but this is hard problem and we don't always get it right. As far as I know, this potential issue has always been there also with previous versions of Bundler versions, it just did not affect your particular case until now.

As I said before, I will try to fix this. You're also welcome to send improvements our way of course. 

Stay safe!

**Comment by doudou** [comment]
@deivid-rodriguez. Not scared :) I appreciate the help (and your collective work on bundler).

**Comment by addersuk** [comment]
Hi, so i've just been investigating the same issue.  I've seen that 2.4.* versions all installed the pre-release gem and when i ran using 2.3.26 the correct production gem was installed.  

This was on a private project on a clean `bundle install` with no lock file.

**Comment by deivid-rodriguez** [comment]
It's just a proof of concept for now, but feel free to experiment with https://github.com/rubygems/rubygems/pull/6246 and see how it works for you.

**Comment by addersuk** [comment]
amazing @deivid-rodriguez, this is working in my situation.

I've manually ran the script using PR https://github.com/rubygems/rubygems/pull/6246 and it picked the same version as Bundler 2.3.26. 

Further details the issue I'm seeing is we have a gem that is using 2 internal gems dependencies hosted on Nexus, one has dry-container with the following in the gemspec

```spec.add_runtime_dependency 'dry-container', '~> 0.7'```

This resulted in 0.11 was being used, but another install gem has stricter versioning with the following in the gemspec

```spec.add_dependency 'dry-container', '~> 0.9.0'```

Meanwhile dependabot has decided to make a PR that has been built as a prerelease and updated this to 

```spec.add_dependency 'dry-container', '>= 0.9', '< 0.12'```

This dependabot prerelease gem is chosen on a fresh bundle install with 2.4* versions.

**Comment by deivid-rodriguez** [comment]
> I've manually ran the script using PR https://github.com/rubygems/rubygems/pull/6246 and it picked the same version as Bundler 2.3.26.

🎉 That's great to hear, I hope to finish that PR soon.

> Further details the issue I'm seeing is we have a gem that is using 2 internal gems dependencies hosted on Nexus, one has dry-container with the following in the gemspec

> `spec.add_runtime_dependency 'dry-container', '~> 0.7'`

> This resulted in 0.11 was being used, but another install gem has stricter versioning with the following in the gemspec

> `spec.add_dependency 'dry-container', '~> 0.9.0'`

🤔 That seems like a bug, since the `~> 0.9.0` requirement is not being respected, right?

> Meanwhile dependabot has decided to make a PR that has been built as a prerelease and updated this to

> `spec.add_dependency 'dry-container', '>= 0.9', '< 0.12'`

I didn't understand this. What does "make a PR that has been built as a prerelease" mean?

**Comment by addersuk** [comment]
> > I've manually ran the script using PR #6246 and it picked the same version as Bundler 2.3.26.
> 
> 🎉 That's great to hear, I hope to finish that PR soon.
> 
> > Further details the issue I'm seeing is we have a gem that is using 2 internal gems dependencies hosted on Nexus, one has dry-container with the following in the gemspec
> 
> > `spec.add_runtime_dependency 'dry-container', '~> 0.7'`
> 
> > This resulted in 0.11 was being used, but another install gem has stricter versioning with the following in the gemspec
> 
> > `spec.add_dependency 'dry-container', '~> 0.9.0'`
> 
> 🤔 That seems like a bug, since the `~> 0.9.0` requirement is not being respected, right?

Sorry to clarify, without this gem it resolved to 0.11, but when this second gem was added, it had to downgrade, but Bundler 2.4.x instead used the prerelease version of privategem 1.2.3.dependabotbundlerdrycontainergte09andlt012  as that had the newer version of the gem instead of the production version privategem.1.3.2

> 
> > Meanwhile dependabot has decided to make a PR that has been built as a prerelease and updated this to
> 
> > `spec.add_dependency 'dry-container', '>= 0.9', '< 0.12'`
> 
> I didn't understand this. What does "make a PR that has been built as a prerelease" mean?

We have automated workflow to build and push a branch gem to a private registry to help integration tests, so we have a gem in our private registry called for example privategem 1.2.3.dependabotbundlerdrycontainergte09andlt012
